src/HappyConnOpener.cc

```diff
@@ -718,6 +727,10 @@ HappyConnOpener::noteSpareAllowance()
     spareWaiting.clear();
     Must(!gotSpareAllowance);
     gotSpareAllowance = true;
+
+    if (ranOutOfTimeOrAttempts())
+        return; // will quit or continue working on prime
```

Comment:
emm
Are you sure?
My sense is that when the HappyConnOpener::noteSpareAllowance scheduled the allowance removed from TheSpareAllowanceGiver (or similar sequence call is executed).

So probably we should move the ranOutOfTimeOrAttempts() before "gotSpareAllowance = true;" but we should not call the "TheSpareAllowanceGiver.jobDroppedAllowance();" inside noteSpareAllowance...

Reply:
> when the HappyConnOpener::noteSpareAllowance scheduled the allowance removed from TheSpareAllowanceGiver

What makes you think that? When noteSpareAllowance is scheduled, the allowance is _created_: The giver calls recordAllowance() which increments the number of allowances given.  Each new allowance delays the next giving event. If we do not call jobDroppedAllowance() nothing will decrement that counter, and others waiting for their allowances will suffer (i.e., will wait longer than necessary).

Each noteSpareAllowance() call must be followed by either jobDroppedAllowance() or jobUsedAllowance(). If we simply quit noteSpareAllowance() then neither will happen.

Reply:
resolved at ae561a10

> and others waiting for their allowances will suffer (i.e., will wait longer than necessary).

As is the code actually the allowance will never released because the spare will be false in swanSong...

Reply:
> As is the code actually the allowance will never released because the spare will be false in swanSong...

I assume that you were referring to earlier branch code. As of ae561a1, swanSong() becomes irrelevant (for this case) because we are dropping the reservation right inside noteSpareAllowance().

I consider this change request addressed. If there is more work to do here, please let me know.